runtime/capsules/src/mctp/driver.rs

```diff
@@ -0,0 +1,355 @@
+            Err(ErrorCode::INVAL)?;
+        }
+        let msg_tag = (arg2 >> 8 & 0xFF) as u8;
+
+        // Receive Request message or send Request message
+        if ((command_num == 1 || command_num == 3) && msg_tag != MCTP_TAG_OWNER)
+            || ((command_num == 2 || command_num == 4) && msg_tag & MCTP_TAG_OWNER != 0)
+        {
+            Err(ErrorCode::INVAL)?;
+        }
+
+        Ok((peer_eid, msg_type, msg_tag))
+    }
+}
+
+impl<'a> SyscallDriver for MCTPDriver<'a> {
+    fn command(
+        &self,
+        command_num: usize,
+        arg1: usize,
+        arg2: usize,
+        process_id: ProcessId,
+    ) -> CommandReturn {
+        let (peer_eid, msg_type, msg_tag) = match self.validate_args(command_num, arg1, arg2) {
+            Ok((peer_eid, msg_type, msg_tag)) => (peer_eid, msg_type, msg_tag),
+            Err(e) => return CommandReturn::failure(e),
+        };
+
+        match command_num {
+            0 => CommandReturn::success(),
+            // 1: Receive Request Message
+            // 2: Receive Response Message
+            1 | 2 => self
+                .apps
+                .enter(process_id, |app, _| {
+                    app.pending_rx = Some(OpContext {
+                        msg_tag,
+                        peer_eid,
+                        msg_type,
+                        op_type: OpType::Rx,
+                    });
+                    CommandReturn::success()
+                })
+                .unwrap_or_else(|err| CommandReturn::failure(err.into())),
+            // 3. Send Request Message
+            // 4: Send Response Message
+            3 | 4 => {
+                let result = self
+                    .apps
+                    .enter(process_id, |app, kernel_data| {
+                        let dest_eid = arg1 as u8;
+                        if app.pending_tx.is_some() {
+                            return Err(ErrorCode::BUSY);
+                        }
+
+                        let res = kernel_data
+                            .get_readonly_processbuffer(ro_allow::MESSAGE_WRITE)
+                            .and_then(|write| {
+                                write.enter(|wpayload| {
+                                    self.kernel_msg_buf.take().map_or(
+                                        Err(ErrorCode::NOMEM),
+                                        |mut kernel_msg_buf| {
+                                            if wpayload.len() > kernel_msg_buf.len() {
+                                                return Err(ErrorCode::SIZE);
+                                            }
+                                            wpayload.copy_to_slice(
+                                                &mut kernel_msg_buf[..wpayload.len()],
+                                            );
+                                            kernel_msg_buf.slice(0..wpayload.len());
+
+                                            match self.sender.send_msg(
```

Comment:
This is getting deeply nested -- maybe break up this `command` function or some parts of it so it is easier to follow along? Some comments probably could help as well.

Reply:
I cleaned send message part of `command` a little and added some comments. This might need further refining once I start testing user apps. Let me know how it looks now.